core/sdk-core/src/main/java/software/amazon/awssdk/core/internal/async/CompressionAsyncRequestBody.java

```diff
@@ -0,0 +1,146 @@
+ */
+@SdkInternalApi
+public class CompressionAsyncRequestBody implements AsyncRequestBody {
+
+    private static final int COMPRESSION_CHUNK_SIZE = 128 * 1024;
+    private final AsyncRequestBody wrapped;
+    private final Compressor compressor;
+
+    private CompressionAsyncRequestBody(DefaultBuilder builder) {
+        Validate.notNull(builder.asyncRequestBody, "wrapped AsyncRequestBody cannot be null");
+        Validate.notNull(builder.compressor, "compressor cannot be null");
+        this.wrapped = builder.asyncRequestBody;
+        this.compressor = builder.compressor;
+    }
+
+    /**
+     * @return Builder instance to construct a {@link CompressionAsyncRequestBody}.
+     */
+    public static Builder builder() {
+        return new DefaultBuilder();
+    }
+
+    public interface Builder extends SdkBuilder<CompressionAsyncRequestBody.Builder, CompressionAsyncRequestBody> {
+
+        /**
+         * Sets the AsyncRequestBody that will be wrapped.
+         * @param asyncRequestBody
+         * @return This builder for method chaining.
+         */
+        Builder asyncRequestBody(AsyncRequestBody asyncRequestBody);
+
+        /**
+         * Sets the compressor to compress the request.
+         * @param compressor
+         * @return This builder for method chaining.
+         */
+        Builder compressor(Compressor compressor);
+    }
+
+    private static final class DefaultBuilder implements Builder {
+
+        private AsyncRequestBody asyncRequestBody;
+        private Compressor compressor;
+
+        @Override
+        public CompressionAsyncRequestBody build() {
+            return new CompressionAsyncRequestBody(this);
+        }
+
+        @Override
+        public Builder asyncRequestBody(AsyncRequestBody asyncRequestBody) {
+            this.asyncRequestBody = asyncRequestBody;
+            return this;
+        }
+
+        @Override
+        public Builder compressor(Compressor compressor) {
+            this.compressor = compressor;
+            return this;
+        }
+    }
+
+    @Override
+    public Optional<Long> contentLength() {
+        return Optional.empty();
+    }
+
+    @Override
+    public String contentType() {
+        return wrapped.contentType();
+    }
+
+    @Override
+    public void subscribe(Subscriber<? super ByteBuffer> s) {
+        Validate.notNull(s, "Subscription MUST NOT be null.");
+
+        ChunkBufferWithUnknownLength chunkBuffer = ChunkBufferWithUnknownLength.builder()
+                                                                               .bufferSize(COMPRESSION_CHUNK_SIZE)
+                                                                               .build();
+
+        wrapped.flatMapIterable(chunkBuffer::bufferAndCreateChunks)
+               .subscribe(new CompressionSubscriber(s, compressor));
+    }
+
+    private static final class CompressionSubscriber implements Subscriber<ByteBuffer> {
+
+        private final Subscriber<? super ByteBuffer> wrapped;
```

Comment:
I realize that the pattern is coming from the ChecksumCalculatingAsyncRequestBody, but the parameter name `wrapped` does not make sense to me here. It's used in the main class for the wrapped AsyncRequestBody, which is appropriate, but here it's confusing. Isn't it the http client (Netty) that's the subscriber?